### DISCUSSION TITLE
How to access URL params in a server component.

### DISCUSSION THREAD

**Discussion by alawi2306** [vercel/next.js#67253] [App Router]
This may have been answered already, but I have had this issue for so long
. 
I need to retrieve a single post based on the URL params, but useParams is a client side hook, and of course you can't use client side components asynchronously. 

There is an open discussion that only deals with the page router:
https://github.com/vercel/next.js/discussions/51818

Therefore here is what I am currently doing:
1. I click on a dynamic route. 
2. Server action renders all of the posts 
3. I then filter the posts from the params getting the required one. This has a O(n) time complexity, whereas a db lookup is O(1).
This therefore results in a 60 on the lighthouse page speed performance metric. 
Does anyone have any idea how to resolve this?

**Comment by serdar-webdev** [comment]
you can use headers function from next js for server components and get all needed headers sent from browser to your server;
if its not enough, you can use middleware to add your custom header with all needed info, and use it via headers;
ex : headers().get('x-url');
